Cards/CardBase.gd

```diff
@@ -18,78 +18,37 @@ class_name CardBase
 @export var damage_to_apply_to_caster: float = 0.0
 @export var status_to_apply_to_target: Array[StatusBase]
 @export var status_to_apply_to_caster: Array[StatusBase]
-@export var affect_all_targets: bool = false
-@export var affect_all_casters: bool = false
+# @export var affect_all_targets: bool = false
+# @export var affect_all_casters: bool = false
+# not needed since we use a list of targets, and the party can also be targets
 @export var amount_of_cards_to_draw: int = 0
 @export var amount_of_cards_to_discard: int = 0
 @export var application_type: Enums.ApplicationType = Enums.ApplicationType.ENEMY_ONLY
 @export var card_title: String = "NULL"
 @export var card_key_art: ImageTexture = null
 @export var card_description: String = "NULL"
 
+var card_effects_data: Array[EffectData] = []
 
-func can_play_card(caster: Entity, target: Entity) -> bool:
-	if caster.get_party_component().can_play_on_entity(application_type, target):
-		return true
-	return false
-
-
-func on_card_play(caster: Entity, target: Entity) -> void:
-	_deal_damage(caster, target)
-	_apply_status(caster, target)
-	_draw_cards()
-	_discard_random_cards()
-	# TODO add other functionality that lots of cards may share (eg: restore AP)
-
+func _ready() -> void:
+	card_effects_data = []
 
-# override in child cards if you want to deal damage in a unique way
-func _deal_damage(caster: Entity, target: Entity) -> void:
-	# damage target
-	if damage_to_apply_to_target != 0.0:
-		_damage_entity(caster, target, damage_to_apply_to_target, affect_all_targets)
-	
-	#damage caster
-	if damage_to_apply_to_caster != 0.0:
-		_damage_entity(caster, caster, damage_to_apply_to_caster, affect_all_casters)
+# Need a parser of some sort to read a plainform card data and translate it to EffectData
+# This will call EffectData.add_effect_data() in a loop for every new effect
```

Comment:
Maybe instead of passing all the data to the EffectData from here, we instead determine what it should be from properties set in the editor. So EffectData would look more like this:
```GDScript
#EffectData.gd
@export var effect: EffectBase = null
@export var application_type: ApplicationType = Enums.ApplicationType.ENEMY_ONLY
@export var value: int = 0
```
So with this, we just set the EffectData array directly from the godot inspector on each CardBase because of the `@export` tag. Then to determine the proper `list_enemy` and `caster` we use `application_type`.

Reply:
When we set from inspector, is that not manual ? how does it work when we are doing this automatically ?
I also don't really like to have an `application_type` honestly.

Reply:
How would setting up EffectData work automatically? I imagine there's always a bit of manual setup for each of the EffectData - otherwise how would you set the `value` or associated `EffectBase`?

For the `application_type`, it's going to be useful when we implement instant-cast cards #21. For an instant-cast card, the `application_type` will be used to determine which party to cast on. Otherwise I'm not sure how we would determine who to cast on since the player isn't selecting a target.

Reply:
The goal of the parser is to translate a card into our infrastructure automatically, so we don't need to manually create each card in the editor. It would work by having a certain structure when writing cards (like the JSON file I mentioned).

I also thought about the `application_type`. The instant-cast card can know who to cast to because of the list of enemies. But it might not know that the list of enemies is the entire enemy team, so that's why the `ENEMIES_ALL` could be useful. But for the logic of applying effects I think it's better to have a list like that.

We could do the following:
- Each card has either a type ALL or SELECT  (the select could be a mono-target, or for example 1 target and the two adjacent to the target, but you would still need to select a main target anyway...). Actually it'll probably be 4, SELECT_ENEMY, SELECT_ALLY, ALL_ENEMY, ALL_ALLY
- Depending on the type: if it's an ALL card, we populate the list with all the party member (depending if it's enemy or ally). If it's a select, the list will get the selected elements (that the player is clicking on / dragging on / whatever the pointing system is).
But ultimately we would keep the list of targets in this system, we just have an enum to know how to fill the list.

Reply:
Is the parser the thing you mentioned in the issue that auto generates the card from a text description? 

That system for targeting may make more sense. If we do it like I was thinking, we would have an application_type and a cast_type that together form the 4 combinations you were thinking of. Consolidating to one is probably better if we don’t need them separate for any reason. Would that still work for cards that target both the enemy and player? (Ex: heal 1 hp to caster, deal 5 to all enemies)

Reply:
After having worked with the inspector to make some cards, it works nicely. The problem of that is that you need to have Godot to make cards. On the other hand, it allows us to not have to write a parser (which could be complicated due to the way Godot writes in resource files, which I'm not exactly sure to understand all the time).

And to answer your question @cheesefrycook the way we can do for cards with multiple effects is to just consider the effect with the most restrictive cast (ie if you have an ALL and a TARGET effect, then the card becomes a TARGET card, you can't auto-cast because it has an effect with a target). And we can easily have multiple effects on the same card now.

Reply:
Basically repeating @cheesefrycook's question - would we be able to affect both player and an enemy with different effects with current setup ? - since on_card_play applies all effects to just the card target ?
EffectData does have both _caster and _target variables, but per my understanding caster is never actually set and target is the same for all effects on the card ?

Reply:
`on_card_play` applies effect to the target. But effects that apply to the player don't need a target, since we only have one player. If you have a heal effect, it will heal the player.
Caster is not currently set because we are waiting for tag system for targeting. And yes target is the same for all effects on the card, because you may only target one entity when a card (and effects that apply to all targets don't need a target).

Reply:
So any effect that would target caster would have to be made individually as one without a target - so if we already had heal with a target for example we would have to make heal without one that just heals the caster ? Probably not a problem, and don't think there will be too many like that I imagine, just wondering 🤔

Reply:
No, the effect can have the caster be both caster and target. If you have an effect that heals self and a target, then the system will use the effect with a targeting to self and the other selected target.

Reply:
I mean if we already had a regular targetted heal that takes target and heals it, that the monsters would use for example and then we wanted to make it target the caster, wouldn't we need to create a new one for it ? Or am I missing something ?

Reply:
The caster is an entity. So you can target the target (ie target yourself).
In practice, we will implement tags that tells you if the target is self or not, so you can know in advance if you need to check for a target. That way we can use the same effect for both targeting someone else and self, but just by changing the tag associated with it (and we also have a tag for teams, so with a different tag you can heal an entire team).

Reply:
In a simpler fashion, currently you can't target both yourself and a target at the same time. That will come later, and that was not the goal here. Currently you can target any entity on the board, but one at a time.